src/nvim/mark_extended.c

```diff
@@ -0,0 +1,1134 @@
+  if (!uhp) {
+    return false;
+  }
+
+  if (kv_size(uhp->uh_extmark) < 1) {
+    return false;
+  }
+  // Check the last action
+  ExtmarkUndoObject object = kv_last(uhp->uh_extmark);
+
+  if (object.type != kColAdjust) {
+    return false;
+  }
+  ColAdjust undo = object.data.col_adjust;
+  bool compactable = false;
+
+  if (!undo.lnum_amount && !lnum_amount) {
+    if (undo.lnum == lnum) {
+      if ((undo.mincol + undo.col_amount) >= mincol) {
+          compactable = true;
+  } } }
+
+  if (!compactable) {
+    return false;
+  }
+
+  undo.col_amount = undo.col_amount + col_amount;
+  ExtmarkUndoObject new_undo = { .type = kColAdjust,
+                                 .data.col_adjust = undo };
+  kv_last(uhp->uh_extmark) = new_undo;
+  return true;
+}
+
+// Save col_adjust info so we can undo/redo
+void u_extmark_col_adjust(buf_T *buf, linenr_T lnum, colnr_T mincol,
+                          long lnum_amount, long col_amount)
+{
+  u_header_T  *uhp = force_get_undo_header(buf);
+  if (!uhp) {
+    return;
+  }
+
+  if (!u_compact_col_adjust(buf, lnum, mincol, lnum_amount, col_amount)) {
+    ColAdjust col_adjust;
+    col_adjust.lnum = lnum;
+    col_adjust.mincol = mincol;
+    col_adjust.lnum_amount = lnum_amount;
+    col_adjust.col_amount = col_amount;
+
+    ExtmarkUndoObject undo = { .type = kColAdjust,
+                               .data.col_adjust = col_adjust };
+
+    kv_push(uhp->uh_extmark, undo);
+  }
+}
+
+// Save col_adjust_delete info so we can undo/redo
+void u_extmark_col_adjust_delete(buf_T *buf, linenr_T lnum,
+                                 colnr_T mincol, colnr_T endcol, int eol)
+{
+  u_header_T  *uhp = force_get_undo_header(buf);
+  if (!uhp) {
+    return;
+  }
+
+  ColAdjustDelete col_adjust_delete;
+  col_adjust_delete.lnum = lnum;
+  col_adjust_delete.mincol = mincol;
+  col_adjust_delete.endcol = endcol;
+  col_adjust_delete.eol = eol;
+
+  ExtmarkUndoObject undo = { .type = kColAdjustDelete,
+                             .data.col_adjust_delete = col_adjust_delete };
+
+  kv_push(uhp->uh_extmark, undo);
+}
+
+// Save adjust info so we can undo/redo
+static void u_extmark_adjust(buf_T * buf, linenr_T line1, linenr_T line2,
+                             long amount, long amount_after)
+{
+  u_header_T  *uhp = force_get_undo_header(buf);
+  if (!uhp) {
+    return;
+  }
+
+  Adjust adjust;
+  adjust.line1 = line1;
+  adjust.line2 = line2;
+  adjust.amount = amount;
+  adjust.amount_after = amount_after;
+
+  ExtmarkUndoObject undo = { .type = kLineAdjust,
+                             .data.adjust = adjust };
+
+  kv_push(uhp->uh_extmark, undo);
+}
+
+// save info to undo/redo a :move
+void u_extmark_move(buf_T *buf, linenr_T line1, linenr_T line2,
+                    linenr_T last_line, linenr_T dest, linenr_T num_lines,
+                    linenr_T extra)
+{
+  u_header_T  *uhp = force_get_undo_header(buf);
+  if (!uhp) {
+    return;
+  }
+
+  AdjustMove move;
```

Comment:
what's the difference between `AdjustMove` and `Adjust` ? Should be a comment near the structs. Not to mention the other `FooAdjust` types...